runtime/lua/vim/lsp.lua

```diff
@@ -0,0 +1,791 @@
+--   name = string | nil;
+--   trace = 'off' | 'messages' | 'verbose' | nil
+--   default_server_callbacks = table | nil;
+--   on_init = function | nil;
+--   init_options = table | nil;
+-- }
+--
+-- - `name` here is only used for logging/debugging.
+-- - `trace` will be forwarded to the client.
+-- - `default_server_callbacks` should be a table of functions which
+-- defines:
+--   - The handlers for notifications. These should be `function(params)`
+--   - A default callback to use for `vim.lsp.buf_request` if one is not
+--   provided at the time of calling `vim.lsp.buf_request`. These should be
+--   `function(err, result)`
+--   - By default, the functions from the module `vim.lsp.builtin_callbacks`
+--   will be used. This parameter can override or extend it those builtin
+--   callbacks.
+--
+-- You can use |vim.lsp.get_client_by_id()| to get the actual client.
+--
+-- NOTE: The client is only available *after* it has been initialized, which
+-- may happen after a small delay (or never if there is an error).
+-- For this reason, you may want to use `on_init` to do any actions once the
+-- client has been initialized.
+--
+-- @return client_id
+function lsp.start_client(conf)
+  assert(type(conf.cmd) == 'string', "conf.cmd must be a string")
+  assert(type(conf.cmd_args) == 'table', "conf.cmd_args must be a table")
+  local offset_encoding = validate_encoding(conf.offset_encoding)
+  -- TODO should I be using this for both notifications and request callbacks
+  -- or separate those?
+  local default_server_callbacks
+  if conf.default_server_callbacks then
+    assert(type(conf.default_server_callbacks) == 'table', "conf.default_server_callbacks must be a table")
+    default_server_callbacks = vim.tbl_extend("keep", conf.default_server_callbacks, builtin_default_server_callbacks)
+  else
+    default_server_callbacks = builtin_default_server_callbacks
+  end
+  -- TODO keep vim.schedule here?
+  for k, v in pairs(default_server_callbacks) do
+    default_server_callbacks[k] = vim.schedule_wrap(v)
+  end
+  local capabilities = conf.capabilities or {}
+  assert(type(capabilities) == 'table', "conf.capabilities must be a table")
+
+  -- There are things sent by the server in the initialize response which
+  -- contains capabilities that would be useful for completion engines, such as
+  -- the character code triggers for completion and code action, so I'll expose this
+  -- for now.
+  if conf.on_init then
+    assert(type(conf.on_init) == 'function', "conf.on_init must be a function")
+  end
+  if conf.on_exit then
+    assert(type(conf.on_exit) == 'function', "conf.on_exit must be a function")
+  end
+  if conf.on_error then
+    assert(type(conf.on_error) == 'function', "conf.on_error must be a function")
+  end
+  if conf.cmd_env then
+    assert(type(conf.cmd_env) == 'table', "conf.cmd_env must be a table")
+  end
+  if conf.cmd_cwd then
+    assert(type(conf.cmd_cwd) == 'string', "conf.cmd_cwd must be a string")
```

Comment:
I think it's useful to make `cmd_cwd` a function which receives buffer as arg and returns a directory which will be used as a cwd for LSP server and treated as a project root. That way inside neovim we could have several instances of LSP servers running for the same filetype but for different project.

Reply:
the `cmd_cwd` is strictly for launching the cmd. This is for cases where the LSP command may expect to be launched from a directory (like if it has relative file pointers). What you're talking about is the `rootUri`, which does raise a very good point which I didn't consider which is configuring multiple of the same LSP for different projects within a single instance of neovim.

The question is whether that is a common enough use-case to warrant a configuration option via `lsp.add_config`. `lsp.add_config` is just for making the common case of launching a single LSP for multiple filetypes easier. If you want a more unique use-case, `lsp.start_client & lsp.attach_to_buffer` is the recommended solution and it is still very easy to use. I think if the documentation explains that clearly and has good examples, then it should be fine.

Also, what you're talking about may be covered by the `workspaceFolder` part of LSP, where a single workspace can have multiple folders.

If you wouldn't mind, @andreypopp  it sounds like you have a specific use-case in mind, so I'd be interested in seeing if you and I could get your existing configuration working with this branch. That would be a good test case. My use-case has been generally more straightforward so far.